File: tests/dump_tokens.py
#!/usr/bin/env python3
"""Dump PyYAML's token stream in the same canonical format as the Rust
`dump_tokens` binary, for differential testing."""

import sys

import yaml

sys.path.insert(0, "tests")


def main():
    path = sys.argv[1]
    with open(path, "rb") as f:
        data = f.read()

    # Same auto-decode as yamllint
    from yamllint.decoder import auto_decode
    content = auto_decode(data)

    loader = yaml.BaseLoader(content)
    try:
        while True:
            token = loader.get_token()
            if token is None:
                break
            name = type(token).__name__.removesuffix("Token")
            value = ""
            if isinstance(token, yaml.DirectiveToken):
                value = f"{token.name} {format_directive(token.value)}"
            elif isinstance(token, (yaml.AliasToken, yaml.AnchorToken)):
                value = token.value
            elif isinstance(token, yaml.TagToken):
                handle, suffix = token.value
                value = f"({handle if handle is not None else 'None'}, {suffix})"
            elif isinstance(token, yaml.ScalarToken):
                style = token.style if token.style is not None else "None"
                value = (
                    f"plain={'true' if token.plain else 'false'} "
                    f"style={style} value={rust_debug(token.value)}"
                )
            s, e = token.start_mark, token.end_mark
            print(
                f"{name} [{s.pointer},{s.line},{s.column}]"
                f"-[{e.pointer},{e.line},{e.column}] {value}".rstrip()
            )
    except yaml.error.MarkedYAMLError as err:
        m = err.problem_mark
        if m is not None:
            print(f"ERROR {err.problem} [{m.pointer},{m.line},{m.column}]")
        else:
            print(f"ERROR {err.problem}")


def format_directive(value):
    if value is None:
        return "None"
    a, b = value
    return f"({a}, {b})"


def rust_debug(s):
    """Format a string like Rust's {:?}."""
    out = ['"']
    for ch in s:
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif ch == "\n":
            out.append("\\n")
        elif ch == "\r":
            out.append("\\r")
        elif ch == "\t":
            out.append("\\t")
        elif ch == "\0":
            out.append("\\0")
        elif ch.isprintable():
            out.append(ch)
        else:
            code = ord(ch)
            out.append(f"\\u{{{code:x}}}")
    out.append('"')
    return "".join(out)


if __name__ == "__main__":
    main()

File: src/bin/dump_tokens.rs
#![forbid(unsafe_code)]

//! Development tool: dump the token stream of a YAML file in a canonical
//! format, for differential testing against PyYAML (see
//! `tests/dump_tokens.py`).

use fastymllint::pyyaml::scanner::Scanner;
use fastymllint::pyyaml::tokens::{DirectiveValue, TokenKind};

fn main() {
    let path = std::env::args().nth(1).expect("usage: dump_tokens FILE");
    let data = std::fs::read(&path).expect("read file");
    let content = fastymllint::decoder::auto_decode(&data).expect("decode");

    let (tokens, error, _buffer) = Scanner::scan_all(&content);
    for t in tokens {
        let (name, value): (&str, String) = match &t.kind {
            TokenKind::StreamStart => ("StreamStart", String::new()),
            TokenKind::StreamEnd => ("StreamEnd", String::new()),
            TokenKind::Directive { name, value } => (
                "Directive",
                match value {
                    DirectiveValue::Yaml(a, b) => format!("{name} ({a}, {b})"),
                    DirectiveValue::Tag(h, p) => format!("{name} ({h}, {p})"),
                    DirectiveValue::Other => format!("{name} None"),
                },
            ),
            TokenKind::DocumentStart => ("DocumentStart", String::new()),
            TokenKind::DocumentEnd => ("DocumentEnd", String::new()),
            TokenKind::BlockSequenceStart => ("BlockSequenceStart", String::new()),
            TokenKind::BlockMappingStart => ("BlockMappingStart", String::new()),
            TokenKind::BlockEnd => ("BlockEnd", String::new()),
            TokenKind::FlowSequenceStart => ("FlowSequenceStart", String::new()),
            TokenKind::FlowMappingStart => ("FlowMappingStart", String::new()),
            TokenKind::FlowSequenceEnd => ("FlowSequenceEnd", String::new()),
            TokenKind::FlowMappingEnd => ("FlowMappingEnd", String::new()),
            TokenKind::Key => ("Key", String::new()),
            TokenKind::Value => ("Value", String::new()),
            TokenKind::BlockEntry => ("BlockEntry", String::new()),
            TokenKind::FlowEntry => ("FlowEntry", String::new()),
            TokenKind::Alias(v) => ("Alias", v.clone()),
            TokenKind::Anchor(v) => ("Anchor", v.clone()),
            TokenKind::Tag { handle, suffix } => (
                "Tag",
                format!("({}, {})", handle.as_deref().unwrap_or("None"), suffix),
            ),
            TokenKind::Scalar {
                value,
                plain,
                style,
            } => (
                "Scalar",
                format!(
                    "plain={} style={} value={:?}",
                    plain,
                    style.map(|c| c.to_string()).unwrap_or("None".to_string()),
                    value
                ),
            ),
        };
        let line = format!(
            "{name} [{},{},{}]-[{},{},{}] {value}",
            t.start_mark.pointer,
            t.start_mark.line,
            t.start_mark.column,
            t.end_mark.pointer,
            t.end_mark.line,
            t.end_mark.column,
        );
        println!("{}", line.trim_end());
    }
    if let Some(e) = error {
        match e.problem_mark {
            Some(m) => println!(
                "ERROR {} [{},{},{}]",
                e.problem, m.pointer, m.line, m.column
            ),
            None => println!("ERROR {}", e.problem),
        }
    }
}
